compiler/src/iree/compiler/Codegen/Common/GPU/GPUTensorTile.cpp

```diff
@@ -31,51 +33,138 @@ using mlir::iree_compiler::IREE::LinalgExt::TilingPatterns;
+    // dimensions of different linalg ops match, which is the current status;
+    // but may not hold true in the long term.
+    tileSizes.resize(op.getLoopIteratorTypes().size());
+
+    if (llvm::all_of(tileSizes, [](int64_t s) { return s == 0; })) {
+      return failure();
+    }
+
+    // Tile the current op and fuse its immediate input operands.
+    FailureOr<scf::SCFTilingResult> tilingResult =
+        tileConsumerAndFuseInputProducer(rewriter, op, tileSizes);
+    if (failed(tilingResult)) {
+      return rewriter.notifyMatchFailure(op, "failed to tile consumer");
+    }
+
+    // Replace the tiled op with replacements.
+    rewriter.replaceOp(op, tilingResult->replacements);
+    filter.replaceLinalgTransformationFilter(rewriter,
+                                             tilingResult->tiledOps.front());
+    return success();
+  }
+
+ private:
+  FailureOr<scf::SCFTilingResult> tileConsumerAndFuseInputProducer(
+      RewriterBase &rewriter, TilingInterface consumer,
+      ArrayRef<int64_t> tileSizes) const {
+    // First tile the current op as the consumer op.
+    auto tilingOptions = scf::SCFTilingOptions().setTileSizes(tileSizes);
+    FailureOr<scf::SCFTilingResult> tilingResult =
+        tileUsingSCFForOp(rewriter, consumer, tilingOptions);
+    if (failed(tilingResult)) {
+      return rewriter.notifyMatchFailure(consumer, "failed to tile consumer");
+    }
+
+    if (!fuseInputProducer) return tilingResult;
+    // If there are no generated loops generated, fusion is immaterial.
+    if (tilingResult->loops.empty()) return tilingResult;
+
+    // Collect immediate input operands that are fusable into the tiled loop.
+    // We have tensor extract slice ops taking slices of the untiled op.
+    //
+    // Note that this excludes init operands for correctness. Input operands are
+    // fine to fuse, at the cost of recomputation though.
+    SmallVector<tensor::ExtractSliceOp> candidates;
+    assert(tilingResult->tiledOps.size() == 1);
+    Operation *tiledOp = tilingResult->tiledOps.front();
+    auto dsOp = dyn_cast<DestinationStyleOpInterface>(tiledOp);
+    if (!dsOp) return tilingResult;
+    for (OpOperand *operand : dsOp.getDpsInputOperands()) {
+      auto sliceOp = operand->get().getDefiningOp<tensor::ExtractSliceOp>();
+      if (!sliceOp) continue;
+      auto linalgOp = sliceOp.getSource().getDefiningOp<linalg::LinalgOp>();
+      if (!linalgOp) continue;
+      // Restrict to fully parallel linalg ops for now for simplicity.
+      auto isParallel = [](utils::IteratorType it) {
+        return linalg::isParallelIterator(it);
+      };
+      if (llvm::all_of(linalgOp.getIteratorTypesArray(), isParallel)) {
+        candidates.push_back(sliceOp);
+      }
+    }
+
+    // Fuse the candidate immeidate operands into the tiled loop.
+    OpBuilder::InsertionGuard guard(rewriter);
+    while (!candidates.empty()) {
+      tensor::ExtractSliceOp sliceOp = candidates.front();
+      candidates.pop_back();
```

Comment:
This looks like it is reading from the front but dropping the last element.
```suggestion
      tensor::ExtractSliceOp sliceOp = candidates.back();
      candidates.pop_back();
```
?

Reply:
Good catch!